hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -325,17 +384,65 @@ public boolean equals(Object o) {
             return false;
         }
         QueryDataType that = (QueryDataType) o;
-        return !isCustomType()
-                ? converter.getId() == that.converter.getId()
-                : objectTypeName.equals(that.objectTypeName)
-                        && objectTypeKind == that.objectTypeKind
-                        && Objects.equals(objectTypeMetadata, that.objectTypeMetadata)
-                        && objectFields.equals(that.objectFields);
+        return !isCustomType() ? converter == that.converter : getDigest().equals(that.getDigest());
     }
 
     @Override
     public String toString() {
-        return getClass().getSimpleName() + " {family=" + getTypeFamily() + "}";
+        return objectTypeName != null ? objectTypeName : NAMES.get(this);
+    }
+
+    public static QueryDataType valueOf(String name) {
+        QueryDataType type = TYPES_BY_NAME.get(name);
+        if (type == null) {
+            throw new IllegalArgumentException("No predefined QueryDataType with name " + name);
+        }
+        return type;
+    }
+
+    public static QueryDataType resolveForConverter(Converter converter) {
+        return TYPES[converter.getId()];
+    }
+
+    public static QueryDataType[] values() {
+        return TYPES.clone();
+    }
+
+    private void computeDigest(StringBuilder sb, Set<String> seen) {
+        escape(sb, objectTypeName);
```

Comment:
digest for builtin types does not work (NPE), and if it would, it would be nonsense and not unique, like null[NONE=null]().
Digest is not used for non-UDT QDTs but it is unlike `RelDataTypeImpl`, so should be documented, maybe also with some asserts

Reply:
59b0e87d